### ISSUE TITLE
maploop stops at final?

### ISSUE THREAD

**Issue by moewew** [plk/biber#269]
Consider
```latex
\documentclass{article}
\usepackage[style=authoryear-comp,uniquename=false]{biblatex}

\DeclareSourcemap{
  \maps[datatype=bibtex]{
    \map[foreach={author, editor}]{
      \step[fieldsource=\regexp{$MAPLOOP}, match=\regexp{Smith,\s+John|Smith,\s+J\.\s*A\.}, final]
      \step[fieldset=\regexp{short$MAPLOOP}, fieldvalue={Smith, John A.}]
    }
    \map[foreach={author, editor}]{
      \step[fieldsource=\regexp{$MAPLOOP}, match=\regexp{Doe,\s+Jane|Doe,\s+J\.\s*B\.}, final]
      \step[fieldset=\regexp{short$MAPLOOP}, fieldvalue={Doe, Jane B.}]
    }
  }
}

\usepackage{filecontents}
\begin{filecontents*}{\jobname.bib}
@article{Smith2000,
  author = {Smith, John},
  title = {Some dubious results},
  journal = {A Great Journal},
  year = {2000},
  volume = {1},
  number = {1}
}
@article{Smith2001,
  author = {Smith, John A.},
  title = {Some more dubious results},
  journal = {A Great Journal},
  year = {2001},
  volume = {1},
  number = {1}
}
@article{Smith2003,
  author = {Smith, J. A.},
  title = {Some dubious results},
  journal = {A Great Journal},
  year = {2003},
  volume = {1},
  number = {1}
}
@collection{Doe2001,
  editor = {Doe, Jane},
  title = {Some more dubious results},
  year = {2001},
}
@collection{Doe2002,
  editor = {Doe, Jane B.},
  title = {Some more dubious results},
  year = {2002},
}
\end{filecontents*}
\addbibresource{\jobname.bib}
\addbibresource{biblatex-examples.bib}

\begin{document}
\parencite{Smith2000,Smith2001,Smith2003,sigfridsson}

\parencite{Doe2001,Doe2002}

\printbibliography
\end{document}
```

The sourcemap compresses the `Smith`s but fails to compress the `Doe`s as expected. It turns out that changing the order in the `foreach` to
```latex
\DeclareSourcemap{
  \maps[datatype=bibtex]{
    \map[foreach={editor, author}]{
      \step[fieldsource=\regexp{$MAPLOOP}, match=\regexp{Smith,\s+John|Smith,\s+J\.\s*A\.}, final]
      \step[fieldset=\regexp{short$MAPLOOP}, fieldvalue={Smith, John A.}]
    }
    \map[foreach={editor, author}]{
      \step[fieldsource=\regexp{$MAPLOOP}, match=\regexp{Doe,\s+Jane|Doe,\s+J\.\s*B\.}, final]
      \step[fieldset=\regexp{short$MAPLOOP}, fieldvalue={Doe, Jane B.}]
    }
  }
}
```
shows the exact opposite effect. That suggests to me that the `final` in a `foreach` `\map` applies to the *entire* looped `\map` operation and not only to each looped `\map` individually. The latter seems more useful to me, though (at least here).

**Comment by plk** [comment]
Can you please try the DEV build? I think I see the issue and this fix should deal with it.

**Comment by moewew** [comment]
Works very well in all tests I ran. Thank you very much.

**Comment by plk** [comment]
Fixed in latest 2.13 release.
